src/WatchDogs.Domain/SuspiciousDealDetector.cs

```diff
@@ -0,0 +1,279 @@
+    private static readonly TimeSpan TimeDifferTolerance = TimeSpan.FromSeconds(1);
+    private readonly IDataLoader _dataLoader;
+    private readonly SuspiciousDealDetectorOptions _suspiciousDealDetectorOptions;
+    private readonly ILogger _logger;
+
+    public ConcurrentDictionary<string, ICurrencyBucket> _currencyTradesPairs = new ConcurrentDictionary<string, ICurrencyBucket>();
+    public SuspiciousDealDetector(IDataLoader dataLoader, IOptions<SuspiciousDealDetectorOptions> options, ILogger logger)
+    {
+        _dataLoader = dataLoader;
+        _suspiciousDealDetectorOptions = options.Value;
+        _logger = logger;
+    }
+
+    public async Task<List<Trade>> LoadDealsAsync()
+    {
+        return await _dataLoader.LoadAllTradesAsync();
+    }
+
+    public async Task<List<List<Trade>>> DetectSuspiciousDealsAsync(List<Trade> trades)
+    {
+        List<List<Trade>> dealsAlreadyFilteredByCurrencyPairAndTimeStamp = new();
+
+        await SortTradesByCurrencyPairsAsync(trades);
+
+        var dealsInBuckets = _currencyTradesPairs.Values;
+
+        Log.Information($"Filtering trades into currency buckets. \n" +
+            $"{dealsInBuckets.Count} buckets found.");
+
+        foreach (var bucket in dealsInBuckets)
+        {
+            var tradesFromOneBucket = bucket.Trades;
+
+            var filteredGroups = GroupTradesByTimestampWithTimeTolerance(tradesFromOneBucket, SuspiciousDealDetector.TimeDifferTolerance);
+
+            dealsAlreadyFilteredByCurrencyPairAndTimeStamp.AddRange(filteredGroups);
+        }
+
+        Log.Information($"Filtering by Timestamp. " +
+            $"{dealsAlreadyFilteredByCurrencyPairAndTimeStamp.Count} buckets remains. {TradesTotalNumber(dealsAlreadyFilteredByCurrencyPairAndTimeStamp)} trades total. \n");
+
+        if(dealsAlreadyFilteredByCurrencyPairAndTimeStamp.Count > 0)
+        {
+            var dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction = TradeActionFilter(dealsAlreadyFilteredByCurrencyPairAndTimeStamp);
+
+            Log.Information($"Filtering by Action. {TradesTotalNumber(dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction)} trades total");
+
+            foreach(var deals in dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction)
+            {
+                var finalFilter = differenceInVolumeToBalanceRatioFilter(deals);
+            }
+
+            Log.Information($"Final filtering by Volume to balance ratio. {TradesTotalNumber(dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction)} trades total");
+
+            dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction.RemoveAll(list => !list.Any());
+
+            SuspiciousDealsLogVisulizer(dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction);
+
+            return dealsAlreadyFilteredByCurrencyPairAndTimeStampAndAction;
+
+        }
+
+        return dealsAlreadyFilteredByCurrencyPairAndTimeStamp;
+    }
+
+    #region Currency Bucket Sort
```

Comment:
I would avoid using regions because they make bloated classes seem more clean (you can make a lot of code disappear just by closing the region). I am not saying don't use them, but be careful with them.

Reply:
I have thought that that is the main duty of regions.